system/mp/mpi_wrapper.cpp

```diff
@@ -0,0 +1,513 @@
+        usleep(WAIT_DELAY);
+    }
+    if (completed)
+        canceled = stat.Is_cancelled();
+    else
+        canceled = !noCancellation;
+
+    return stat;
+}
+
+MPI::Status hasIncomingData(int sourceRank, int mptag, MPI::Comm& comm,
+        bool& incomingMessage, bool& error, bool& canceled, unsigned timeout, CommData *commData)
+{
+    _TF("hasIncomingData", sourceRank, mptag, incomingMessage, error, timeout);
+
+    CTimeMon tm(timeout);
+    MPI::Status stat;
+
+    int flag;
+    unsigned remaining;
+    bool noCancellation;
+
+    while ((noCancellation = commData->lockFromCancellation()) && !(incomingMessage || error || (timeout !=0 && tm.timedout(&remaining))))
+    {
+        incomingMessage = comm.Iprobe(sourceRank, mptag, stat);
+        commData->releaseCancellationLock();
+        if (timeout == 0) break;
+        usleep(PROBING_DELAY);
+    }
+    canceled = !noCancellation;
+    return stat;
+}
+
+//----------------------------------------------------------------------------//
+
+/** See mpi_wrapper.hpp header file for function descriptions of the following **/
+
+bool hpcc_mpi::hasIncomingMessage(rank_t &sourceRank, mptag_t &mptag, MPI::Comm& comm)
+{
+    _TF("hasIncomingMessage", sourceRank, mptag);
+    bool incomingMessage = false; bool error = false; bool canceled = false;
+    int source = getRank(sourceRank);
+    int tag = getTag(mptag);
+    CommData* tmpCommData = new CommData(source, tag, comm);
+    MPI::Status stat = hasIncomingData(source, tag, comm, incomingMessage, error, canceled, 0, tmpCommData);
+    if (incomingMessage)
+    {
+        sourceRank = stat.Get_source();
+        mptag = getTag(stat.Get_tag());
+    }
+
+    return incomingMessage;
+}
+
+bool cancelComm(CommData* commData)
+{
+    bool ret = true;
+    if (commData)
+    {
+        MPI::Status stat;
+        //TODO: try catch for error
+        bool completed;
+        completed = commData->request.Test(stat);
+        if (!completed) {
+            commData->request.Cancel();
+            commData->request.Free();
+        }
+    }
+
+    return ret;
+}
+
+bool hpcc_mpi::cancelComm(bool send, rank_t rank, mptag_t mptag, MPI::Comm& comm)
+{
+    _TF("cancelComm", send, rank, mptag);
+    int r = getRank(rank);
+    int tag = getTag(mptag);
+    std::vector<CommData*> commDataList = popCommData(send, r, tag, comm);
+    bool success = true;
+    for(int i=0; i<commDataList.size(); i++)
+    {
+        commDataList[i]->notifyCancellation();              // In case sendData and readData functions still in progress we want to let
+                                                            // them know that we are about to screw up their plans
+        success = success && cancelComm(commDataList[i]);   // If we managed to cancel everything then the cancellation was successful
+        usleep(WAIT_DELAY);                                 // Wait for a short while for active threads to exit send/recv function
```

Comment:
I see the comment, but don't understand why delay here really? We don't have one in the MP impl. afair.
NB: whenever there's a sleep in a loop, we need to think about refactoring. It should be possible to avoid completely, by using tracking objects that are waiting with semaphores and signalling.

Reply:
Yeah I think we should be able to get rid of that. I was thinking of a certain race condition when I put this, but cant remember what it could be now.